### PR TITLE
Implement `MethodImplOptions.AggressiveInlining` flag - fixes #1637

### LINKED ISSUE #1637: The F# compiler ignores `MethodImplOptions.AggressiveInlining` flag
The F# compiler ignores the `System.Runtime.CompilerServices.MethodImplAttribute` applied to methods/functions if the `MethodImplOptions.AggressiveInlining` flag is set and consequently doesn't emit the required IL metadata ("aggressiveinlining" method flag = 0x0100).
#### Descriptive Example

For example, compiling the following F# code

``` fsharp
[<MethodImpl(MethodImplOptions.AggressiveInlining)>]
let inc x = x + 1
```

yields the following IL

```
.method public static int32  inc(int32 x) cil managed
{
    .maxstack  8
    IL_0000:  nop
    IL_0001:  ldarg.0
    IL_0002:  ldc.i4.1
    IL_0003:  add
    IL_0004:  ret
}
```

In contrast, the C# compiler translates

``` csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static int Inc(int x) => x + 1;
```

to

```
.method public hidebysig static int32  Inc(int32 x) cil managed aggressiveinlining
{     
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  ldc.i4.1
    IL_0002:  add
    IL_0003:  ret
}
```

generating the required `aggressiveinling` flag.
#### Further Experiments

The same behavior can be observed for `MethodImplOptions.AggressiveInlining` flags applied to methods or static methods of class types. Strangely enough, if applied to indexer `get`/`set` methods, the F# compiler doesn't ignore the attribute, but treats it like a custom attribute ([instead of a pseudo custom attribute](http://weblog.ikvm.net/2008/11/25/PseudoCustomAttributes.aspx)). For instance,

``` fsharp
type Dummy =
    member self.Item
        with [<MethodImpl(MethodImplOptions.AggressiveInlining)>] get x = x + 1
```

becomes

```
.method public hidebysig specialname instance int32 get_Item(int32 x) cil managed
{
    .custom instance void [mscorlib]System.Runtime.CompilerServices.MethodImplAttribute::.ctor(valuetype [mscorlib]System.Runtime.CompilerServices.MethodImplOptions) = ( 01 00 00 01 00 00 00 00 ) 
    .maxstack  8
    IL_0000:  nop
    IL_0001:  ldarg.1
    IL_0002:  ldc.i4.1
    IL_0003:  add
    IL_0004:  ret
}
```
#### Known workarounds

None.
#### Related information

According to @kvb, `PreserveSig`, `Synchronized`, and `NoInlining` flags seem to be respected by the compiler, see [this StackOverflow question](http://stackoverflow.com/questions/40104687/applying-methodimploptions-aggressiveinlining-to-f-functions).

Test environment:
- Windows 10 Pro Build (14393.321)
- .NET 4.6
- F# 4.0
- VisualStudio 2015

### PR DESCRIPTION
**Pull request by forki**
- [x] - implement #1637 
- [x] - add tests

### CODE REVIEW

**Comment by forki** [comment]
@dsyme @KevinRansom @kbattocchi  I tried to find all places that were touched by NoInline but it's not working yet. Any ideas?

File: src/fsharp/tast.fs [hunk 1]
```diff
@@ -123,6 +125,7 @@ type ValFlags(flags:int64) =
                      (match inlineInfo with
                                         | ValInline.PseudoVal ->             0b0000000000000000000L
                                         | ValInline.Always ->                0b0000000000000010000L
+                                        | ValInline.Aggressive ->            0b0100L
```
Comment (dsyme) [inline src/fsharp/tast.fs]: This is not right - 0b0100L is already used by ``NormalVal`` above

File: src/fsharp/tast.fs [hunk 2]
```diff
@@ -62,6 +62,8 @@ type ValInline =
     | PseudoVal
     /// Indicates the value is inlined but the .NET IL code for the function still exists, e.g. to satisfy interfaces on objects, but that it is also always inlined 
     | Always
+    /// Indicates the value will be inlined be inlined by the .NET runtime
+    | Aggressive
```
Comment (dsyme) [inline src/fsharp/tast.fs]: I'm not convinced we need to change the TAST (or Pickle) at all?

I think all you need to do is use the TAST value ``ValInline.Never`` if the attribute is present.

**Comment by dsyme** [comment]
@forki This is now ready but needs a test.  Adding a code generation baseline test would make sense. Would you like to do that? thanks!

**Comment by forki** [comment]
Sorry in foreseeable future I won't have time to work on it. Cutting all F#
OSS work down to minimum. Only stuff that is directly related to work
blockers will survive.

Am 23.06.2017 5:37 nachm. schrieb "Don Syme" <notifications@github.com>:

**Comment by dsyme** [comment]
> Sorry in foreseeable future I won't have time to work on it. 

Ok, I can add this one

**Comment by dsyme** [comment]
OK, test added, this is ready once green

**Pull request merged**